src/ckeditor.tsx

```diff
@@ -239,26 +299,105 @@ export default class CKEditor<TEditor extends Editor> extends React.Component<Pr
 		} );
 	}
 
+	private getStateDiff( previousState: Record<string, unknown>, newState: Record<string, unknown> ):
+		{ addedKeys: Array<string>; removedKeys: Array<string>; modifiedKeys: Array<string> } {
+		const previousStateKeys = Object.keys( previousState );
+		const newStateKeys = Object.keys( newState );
+
+		return {
+			addedKeys: newStateKeys.filter( key => previousStateKeys.indexOf( key ) === -1 ),
+			removedKeys: previousStateKeys.filter( key => newStateKeys.indexOf( key ) === -1 ),
+			modifiedKeys: previousStateKeys.filter( key => newStateKeys.indexOf( key ) !== -1 &&
+				JSON.stringify( previousState[ key ] ) !== JSON.stringify( newState[ key ] ) )
+		};
+	}
+
+	private _handleRoots = (
+		nextProps: any,
+		newRoots: Array<string>,
+		removedRoots: Array<string>,
+		modifiedRoots: Array<string>
+	) => {
+		const editor = this.editor as MultiRootEditor;
+
+		newRoots.forEach( root => {
+			editor.addRoot( root, {
+				data: nextProps.data[ root ] || '',
+				attributes: nextProps.attributes?.[ root ] || {},
+				isUndoable: true
+			} );
+		} );
+
+		removedRoots.forEach( root => {
+			editor.detachRoot( root, true );
+		} );
+
+		for ( const modifiedRoot of modifiedRoots ) {
+			if ( editor.data.get( { rootName: modifiedRoot } ) !== nextProps.data[ modifiedRoot ] ) {
+				editor!.data.set( nextProps.data! );
+
+				break;
+			}
+		}
+	};
+
+	private _handleRootsAttributes = (
+		writer: any,
+		nextProps: any,
+		newRootsAttributes: Array<string>,
+		removedRootsAttributes: Array<string>,
+		modifiedRootsAttributes: Array<string>
+	) => {
+		[ ...newRootsAttributes, ...modifiedRootsAttributes ].forEach( root => {
+			writer.setAttributes( nextProps.attributes[ root ], this.editor!.model.document.getRoot( root ) );
+		} );
+
+		removedRootsAttributes.forEach( root => {
+			writer.clearAttributes( this.editor!.model.document.getRoot( root ) );
+		} );
```

Comment:
Is this supposed to clear all attributes or remove only some of them?

Reply:
It is called only when someone removes the whole key from the attributes object. It's a very rare situation.
@scofalik I remember that we were talking about it. Do you still confirm that we can keep it as it is or maybe you want to rewrite it in some way?